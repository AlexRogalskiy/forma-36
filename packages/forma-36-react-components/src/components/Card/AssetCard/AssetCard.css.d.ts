// This file is automatically generated by typings-for-css-modules.
// Please do not change this file!
export interface IAssetCardCss {
  'AssetCard': string;
  'AssetCard__asset': string;
  'AssetCard__header': string;
  'AssetCard__header__dropdown-button': string;
}

export const locals: IAssetCardCss;
