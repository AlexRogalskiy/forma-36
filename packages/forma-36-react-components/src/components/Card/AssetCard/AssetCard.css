@import 'settings/colors';
@import 'settings/typography';
@import 'settings/transitions';

.AssetCard {
  display: inline-block;
  padding: 0px;
  min-width: calc(1rem * (240 / var(--rem-base)));
  min-height: calc(1rem * (80 / var(--rem-base)));
}

.AssetCard__header {
  padding: calc(1rem * (14 / var(--rem-base)))
    calc(1rem * (19 / var(--rem-base)));
  display: flex;
  align-items: center;
  border-bottom: calc(1rem * (1 / var(--rem-base))) solid
    var(--color-element-mid);
}

.AssetCard__asset {
  width: 100%;
  height: 100%;
}

.AssetCard__content {
  display: flex;
  align-items: center;
  justify-content: center;
  max-height: calc(1rem * (240 / var(--rem-base)));
  min-width: calc(1rem * (240 / var(--rem-base)));
}

.AssetCard__header__dropdown-button {
  border: none;
  background: transparent;
  display: flex;
  padding: 0px;
  cursor: pointer;
  margin-left: calc(1rem * (10 / var(--rem-base)));
}

.AssetCard__header__dropdown-button svg {
  /* This needs to be set for IE11 to show the icon */
  width: 18px;
}
