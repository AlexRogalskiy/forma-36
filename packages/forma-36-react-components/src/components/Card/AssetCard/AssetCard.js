import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import CardLoading from '../CardLoading';
import Dropdown from '../../Dropdown/Dropdown';
import Asset from '../../Asset';
import Icon from '../../Icon/Icon';
import TabFocusTrap from '../../TabFocusTrap';
import Tag from '../../Tag';
import styles from './AssetCard.css';

class AssetCard extends React.Component {
  static propTypes = {
    extraClassNames: PropTypes.string,
    src: PropTypes.string,
    isLoading: PropTypes.bool,
    type: PropTypes.string,
    title: PropTypes.string,
    dropdownListElements: PropTypes.node,
    status: PropTypes.oneOf(['archived', 'changed', 'draft', 'published']),
    testId: PropTypes.string,
    height: PropTypes.number,
  };

  static defaultProps = {
    extraClassNames: undefined,
    dropdownListElements: undefined,
    isLoading: false,
    src: undefined,
    title: undefined,
    status: undefined,
    type: undefined,
    testId: 'cf-ui-asset-card',
    height: undefined,
  };

  state = {
    isOpen: false,
  };

  renderDropdownListElements = dropdownListElements => (
    <Dropdown
      isOpen={this.state.isOpen}
      onClose={() => this.setState({ isOpen: false })}
      position="bottom-right"
      toggleElement={
        <button
          className={styles['AssetCard__header__dropdown-button']}
          onClick={() => this.setState({ isOpen: !this.state.isOpen })}
        >
          <TabFocusTrap>
            <Icon icon="MoreHorizontalTrimmed" color="secondary" />
          </TabFocusTrap>
        </button>
      }
    >
      {dropdownListElements}
    </Dropdown>
  );

  renderStatus = status => {
    let label;
    let type;

    switch (status) {
      case 'archived':
        label = 'archived';
        type = 'negative';
        break;

      case 'changed':
        label = 'changed';
        type = 'primary';
        break;

      case 'published':
        label = 'published';
        type = 'positive';
        break;

      default:
        label = 'draft';
        type = 'warning';
    }

    return (
      <Tag tagType={type} style={{ marginLeft: 'auto' }}>
        {label}
      </Tag>
    );
  };

  render() {
    const {
      extraClassNames,
      src,
      type,
      title,
      status,
      isLoading,
      dropdownListElements,
      testId,
      height,
      ...otherProps
    } = this.props;

    const classNames = cn(styles.AssetCard, extraClassNames);
    return (
      <CardLoading
        isLoading={isLoading}
        extraClassNames={classNames}
        title={title}
        testId={testId}
        height={height}
        {...otherProps}
      >
        <div className={styles['AssetCard__inner-wrapper']}>
          <div className={styles.AssetCard__header}>
            {status && this.renderStatus(status)}
            {dropdownListElements &&
              this.renderDropdownListElements(dropdownListElements)}
          </div>
          <div className={styles.AssetCard__content}>
            <Asset
              extraClassNames={styles.AssetCard__asset}
              src={src}
              title={title}
              type={type}
            />
          </div>
        </div>
      </CardLoading>
    );
  }
}

export default AssetCard;
