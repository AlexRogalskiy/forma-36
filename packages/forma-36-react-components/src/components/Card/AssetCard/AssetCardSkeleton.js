import React from 'react';
import SkeletonContainer from '../../Skeleton/SkeletonContainer';
import SkeletonBodyText from '../../Skeleton/SkeletonBodyText';
import SkeletonImage from '../../Skeleton/SkeletonImage';

const AssetCardSkeleton = () => (
  <SkeletonContainer svgWidth={240} clipId="f36-asset-card-skeleton">
    <SkeletonImage offsetLeft={85} offsetTop={100} />
    <SkeletonBodyText
      offsetLeft={70}
      offsetTop={190}
      numberOfLines={1}
      width={100}
    />
  </SkeletonContainer>
);

export default AssetCardSkeleton;
