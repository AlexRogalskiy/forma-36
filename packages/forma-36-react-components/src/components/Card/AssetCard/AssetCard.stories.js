import React from 'react';
import { storiesOf } from '@storybook/react';
import { text, selectV2, boolean } from '@storybook/addon-knobs/react';
import { action } from '@storybook/addon-actions';
import { host } from 'storybook-host';
import { withInfo } from '@storybook/addon-info';

import AssetCard from './AssetCard';
import DropdownList from './../../Dropdown/DropdownList';
import DropdownListItem from './../../Dropdown/DropdownListItem';
import { types } from '../../Asset/Asset';

storiesOf('Components|Card/AssetCard', module)
  .addDecorator(
    host({
      align: 'center middle',
      cropMarks: false,
    }),
  )
  .add(
    'default',
    withInfo()(() => (
      <AssetCard
        extraClassNames={text('Extra Class Names', '')}
        status={selectV2(
          'Status',
          {
            Draft: 'draft',
            Changed: 'changed',
            Published: 'published',
            Archived: 'archived',
          },
          'published',
        )}
        type={selectV2('Asset Type', types, 'archive')}
        isLoading={boolean('Is Loading', false)}
        src={text('Source', 'https://placekitten.com/200/300')}
        title={text('Title', 'Image of a cat')}
      />
    )),
  )
  .add(
    'with dropdownListElements',
    withInfo()(() => (
      <AssetCard
        extraClassNames={text('Extra Class Names', '')}
        status={selectV2(
          'Status',
          {
            Draft: 'draft',
            Changed: 'changed',
            Published: 'published',
            Archived: 'archived',
          },
          'published',
        )}
        type={selectV2('Asset Type', types, 'image')}
        isLoading={boolean('Is Loading', false)}
        src={text('Source', 'https://placekitten.com/200/300')}
        title={text('Title', 'Image of a cat')}
        dropdownListElements={
          <span>
            <DropdownList>
              <DropdownListItem isTitle>Actions</DropdownListItem>
              <DropdownListItem onClick={action('Edit onClick')}>
                Edit
              </DropdownListItem>
              <DropdownListItem onClick={action('Download onClick')}>
                Download
              </DropdownListItem>
              <DropdownListItem onClick={action('Remove onClick')}>
                Remove
              </DropdownListItem>
            </DropdownList>
            <DropdownList border="top">
              <DropdownListItem isTitle>File info</DropdownListItem>
              <DropdownListItem>Asset name</DropdownListItem>
              <DropdownListItem>Asset type</DropdownListItem>
              <DropdownListItem>Asset size</DropdownListItem>
              <DropdownListItem>Asset resolution</DropdownListItem>
            </DropdownList>
          </span>
        }
      />
    )),
  );
