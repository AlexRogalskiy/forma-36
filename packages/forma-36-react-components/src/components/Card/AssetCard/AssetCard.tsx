import React, { Component } from 'react';
import cn from 'classnames';
import Card from '../Card';
import CardActions from './../CardActions';
import Asset from '../../Asset';
import { AssetType } from '../../Asset/Asset';
import Tag, { TagType } from '../../Tag/Tag';
import AssetCardSkeleton from './AssetCardSkeleton';
const styles = require('./AssetCard.css');

export type AssetState = 'archived' | 'changed' | 'draft' | 'published';

export type AssetCardProps = {
  src: string;
  title: string;
  className?: string;
  isLoading?: boolean;
  dropdownListElements?: React.ReactElement;
  status?: AssetState;
  testId?: string;
  type?: AssetType;
} & typeof defaultProps;

const defaultProps = {
  isLoading: false,
  testId: 'cf-ui-asset-card',
};

export class AssetCard extends Component<AssetCardProps> {
  static defaultProps = defaultProps;

  renderStatus = (status: AssetState) => {
    let label;
    let type: TagType | null = null;

    switch (status) {
      case 'archived':
        label = 'archived';
        type = 'negative';
        break;

      case 'changed':
        label = 'changed';
        type = 'primary';
        break;

      case 'published':
        label = 'published';
        type = 'positive';
        break;

      default:
        label = 'draft';
        type = 'warning';
    }

    return (
      <Tag tagType={type} style={{ marginLeft: 'auto' }}>
        {label}
      </Tag>
    );
  };

  render() {
    const {
      className,
      src,
      type,
      title,
      status,
      isLoading,
      dropdownListElements,
      testId,
      ...otherProps
    } = this.props;

    const classNames = cn(styles['AssetCard'], className);
    return (
      <Card
        className={classNames}
        title={title}
        testId={testId}
        {...otherProps}
      >
        {isLoading ? (
          <AssetCardSkeleton />
        ) : (
          <div className={styles['AssetCard__inner-wrapper']}>
            <div className={styles['AssetCard__header']}>
              {status && this.renderStatus(status)}
              {dropdownListElements && (
                <CardActions className={styles['AssetCard__actions']}>
                  {dropdownListElements}
                </CardActions>
              )}
            </div>
            <div className={styles['AssetCard__content']}>
              <Asset
                className={styles['AssetCard__asset']}
                src={src}
                title={title}
                type={type}
              />
            </div>
          </div>
        )}
      </Card>
    );
  }
}

export default AssetCard;
