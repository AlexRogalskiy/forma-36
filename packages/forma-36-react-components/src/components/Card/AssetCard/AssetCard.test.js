import React from 'react';
import { shallow, mount } from 'enzyme';
import { axe } from 'jest-axe';
import AssetCard from './AssetCard';

it('renders the component', () => {
  const output = shallow(<AssetCard>AssetCard</AssetCard>);

  expect(output).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const output = shallow(
    <AssetCard
      extraClassNames="my-extra-class"
      src="http://placekitten.com/200/300"
      title="picture of a cat"
    />,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component in loading state', () => {
  const output = shallow(
    <AssetCard
      isLoading
      src="http://placekitten.com/200/300"
      title="picture of a cat"
    />,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component with status', () => {
  const output = shallow(
    <AssetCard
      status="archived"
      src="http://placekitten.com/200/300"
      title="picture of a cat"
    />,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component with actions', () => {
  const output = shallow(
    <AssetCard
      actions={[
        {
          title: 'delete',
          onClick: () => {},
        },
      ]}
      src="http://placekitten.com/200/300"
      title="picture of a cat"
    />,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component without actions', () => {
  const output = shallow(
    <AssetCard src="http://placekitten.com/200/300" title="picture of a cat" />,
  );

  expect(output.find('Dropdown').length).toEqual(0);
  expect(output).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const output = mount(
    <AssetCard src="http://placekitten.com/200/300" title="picture of a cat" />,
  ).html();
  const results = await axe(output);

  expect(results).toHaveNoViolations();
});
