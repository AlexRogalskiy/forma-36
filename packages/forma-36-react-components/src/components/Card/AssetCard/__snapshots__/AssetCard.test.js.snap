// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<CardLoading
  extraClassNames="AssetCard"
  isLoading={false}
  testId="cf-ui-asset-card"
  title="Untitled"
>
  <div
    className="AssetCard__inner-wrapper"
  >
    <div
      className="AssetCard__header"
    />
    <Asset
      extraClassNames="AssetCard__asset"
      type="image"
    />
  </div>
</CardLoading>
`;

exports[`renders the component in loading state 1`] = `
<CardLoading
  extraClassNames="AssetCard"
  isLoading={true}
  testId="cf-ui-asset-card"
  title="picture of a cat"
>
  <div
    className="AssetCard__inner-wrapper"
  >
    <div
      className="AssetCard__header"
    />
    <Asset
      extraClassNames="AssetCard__asset"
      src="http://placekitten.com/200/300"
      title="picture of a cat"
      type="image"
    />
  </div>
</CardLoading>
`;

exports[`renders the component with actions 1`] = `
<CardLoading
  actions={
    Array [
      Object {
        "onClick": [Function],
        "title": "delete",
      },
    ]
  }
  extraClassNames="AssetCard"
  isLoading={false}
  testId="cf-ui-asset-card"
  title="picture of a cat"
>
  <div
    className="AssetCard__inner-wrapper"
  >
    <div
      className="AssetCard__header"
    />
    <Asset
      extraClassNames="AssetCard__asset"
      src="http://placekitten.com/200/300"
      title="picture of a cat"
      type="image"
    />
  </div>
</CardLoading>
`;

exports[`renders the component with an additional class name 1`] = `
<CardLoading
  extraClassNames="AssetCard my-extra-class"
  isLoading={false}
  testId="cf-ui-asset-card"
  title="picture of a cat"
>
  <div
    className="AssetCard__inner-wrapper"
  >
    <div
      className="AssetCard__header"
    />
    <Asset
      extraClassNames="AssetCard__asset"
      src="http://placekitten.com/200/300"
      title="picture of a cat"
      type="image"
    />
  </div>
</CardLoading>
`;

exports[`renders the component with status 1`] = `
<CardLoading
  extraClassNames="AssetCard"
  isLoading={false}
  testId="cf-ui-asset-card"
  title="picture of a cat"
>
  <div
    className="AssetCard__inner-wrapper"
  >
    <div
      className="AssetCard__header"
    >
      <Tag
        style={
          Object {
            "marginLeft": "auto",
          }
        }
        tagType="negative"
        testId="cf-ui-tag"
      >
        archived
      </Tag>
    </div>
    <Asset
      extraClassNames="AssetCard__asset"
      src="http://placekitten.com/200/300"
      title="picture of a cat"
      type="image"
    />
  </div>
</CardLoading>
`;

exports[`renders the component without actions 1`] = `
<CardLoading
  extraClassNames="AssetCard"
  isLoading={false}
  testId="cf-ui-asset-card"
  title="picture of a cat"
>
  <div
    className="AssetCard__inner-wrapper"
  >
    <div
      className="AssetCard__header"
    />
    <Asset
      extraClassNames="AssetCard__asset"
      src="http://placekitten.com/200/300"
      title="picture of a cat"
      type="image"
    />
  </div>
</CardLoading>
`;
